### PR TITLE
Saving the serialized context to a file

### PR DESCRIPTION
**Pull request by piotr-now**
Signed-off-by: Piotr Nowicki <piotr.nowicki@arm.com>

## Description
This PR adds in the ssl_server2 and ssl_client2 programs the ability to save a serialized context to a file as base64 code. 

## Status
**READY**

## Requires Backporting
NO

### CODE REVIEW

File: programs/ssl/ssl_client2.c [hunk 1]
```diff
@@ -516,6 +523,9 @@ struct options
                                  * during renegotiation                     */
     const char *cid_val;        /* the CID to use for incoming messages     */
     int serialize;              /* serialize/deserialize connection         */
+    const char *serial_to_file; /* the file to write a serialized connection
+                                 * in the form of base64 code ( serialize
```
Comment (Patater) [inline programs/ssl/ssl_client2.c:527]: Style: don't need spaces after `(` in comments.

File: programs/ssl/ssl_client2.c [hunk 2]
```diff
@@ -363,9 +365,14 @@ int main( void )
 
 #if defined(MBEDTLS_SSL_CONTEXT_SERIALIZATION)
 #define USAGE_SERIALIZATION \
-    "    serialize=%%d        default: 0 (do not serialize/deserialize)\n" \
-    "                        options: 1 (serialize)\n"                    \
-    "                                 2 (serialize with re-initialization)\n"
+    "    serialize=%%d        default: 0 (do not serialize/deserialize)\n"     \
+    "                        options: 1 (serialize)\n"                         \
+    "                                 2 (serialize with re-initialization)\n"  \
+    "    serial_to_file=%%s   The file path to write a serialized connection\n"\
```
Comment (Patater) [inline programs/ssl/ssl_client2.c:371]: Could we rename this to "context_file"? "Serial to file" doesn't make very much sense in English.

File: programs/ssl/ssl_client2.c [hunk 3]
```diff
@@ -3102,10 +3117,56 @@ int main( int argc, char *argv[] )
...
+
+            mbedtls_printf( "ok\n" );
+            fclose( b64_file );
+        }
+
         if( opt.serialize == 1 )
         {
             /* nothing to do here, done by context_save() already */
-            mbedtls_printf( "  . Context has been reset... ok" );
+            mbedtls_printf( "  . Context has been reset... ok\n" );
```
Comment (Patater) [inline programs/ssl/ssl_client2.c:3169]: This looks like an unrelated change. Is it needed? If so, could we do it in its own commit?

File: programs/ssl/ssl_client2.c (hunk 2 above)
Comment (piotr-now) [reply programs/ssl/ssl_client2.c]: Yes, I know. I wanted to use an abbreviation instead of a full word like for example 'renego' instead of 'renegotiation' :). I changed to 'context_file'

File: programs/ssl/ssl_client2.c (hunk 3 above)
Comment (piotr-now) [reply programs/ssl/ssl_client2.c:3173]: Rather needed. Without it the log looked like this:
```
  . Serializing live connection... ok
  . Context has been reset... ok  . Deserializing connection... ok
  . Closing the connection... done
```
I put it to separate commit

Comment (piotr-now) [reply programs/ssl/ssl_client2.c:3173]: Added missing mbedtls_free to fix CI failure

**Comment by piotr-now** [comment]
Added missing mbedtls_free to fix CI failure

File: programs/ssl/ssl_client2.c (hunk 3 above)
Comment (Patater) [reply programs/ssl/ssl_client2.c:3173]: Thanks

File: tests/ssl-opt.sh [hunk 4]
```diff
@@ -1628,6 +1628,16 @@ run_test    "Context serialization, re-init, both serialize, with CID" \
...
 
+requires_config_enabled MBEDTLS_SSL_CONTEXT_SERIALIZATION
+run_test    "Saving the serialized context to a file" \
+            "$P_SRV dtls=1 serialize=1 context_file=context_srv.txt" \
+            "$P_CLI dtls=1 serialize=1 context_file=context_cli.txt" \
+            0 \
+            -s "Save serialized context to a file... ok" \
+            -c "Save serialized context to a file... ok"
+rm -f context_srv.txt
+rm -f context_cli.txt
```
Comment (Patater) [inline tests/ssl-opt.sh:1639]: Could we move this to the `cleanup()` function instead?

Comment (piotr-now) [reply tests/ssl-opt.sh:1641]: This should be copied to cleanup but not moved. It seems that cleanup is not triggered when the test completes normally but only when certain signals from the system occur:
`trap cleanup INT TERM HUP`

Comment (piotr-now) [reply tests/ssl-opt.sh:1641]: I added it to cleanup

**Comment by Patater** [comment]
CI passing, other than Mbed OS known issue which is being addressed separately. Good to go.
